docs/articles/immutability-in-js.md

```diff
@@ -0,0 +1,108 @@
+
+### Object.assign
+
+We want to add a name property in our object.
+
+https://gist.github.com/frinyvonnick/a281f15288d3ab90ce5354f01df52ccf
+
+We can do it with `Object.assign` and a little trick. Basically it copies all properties of an object to another one, thus it mutates the target object. So we use a small trick, by passing an empty object at first parameter we create a new reference.
+
+https://gist.github.com/frinyvonnick/98878e45d6a8ccb4618238dbec175ae9
+
+So we now have a new object with our new name property and a commonNames property remaining unchanged. With this method you can create/overwrite multiple properties at the same time.
+
+### Array.concat
+
+Now let's do it with an array. We have an array and we want to add two new elements in an immutable way.
+
+https://gist.github.com/frinyvonnick/461d4dbad6bff965a60b734f86619faa
+
+`Array.concat` doesn't mutate our array in contrary of `Array.push` an return a new array without using trick like we did with `Object.assign`.
+
+https://gist.github.com/frinyvonnick/3959e7972688a41055671bceb6228056
+
+This method is flexible. It takes as many elements as you want and they can be values or arrays.
+
+### Object.freeze
+
+`Object.freeze` isn't really familiar. It let you make an object immutable ! It prevents every type of mutation (creation, modification, deletion) included using setter.
+
+https://gist.github.com/frinyvonnick/d1dd1cbf2924b2428b07713d1ad2adf7
+
+We will try to delete name property after object has been frozen.
+
+https://gist.github.com/frinyvonnick/acc86d1c2fe14c23cd6363946cc253da
+
+Reallocation isn't necessary since the object passed as parameter is made immutable by `Object.freeze`. This methode has two modes available:
+
+- In non-strict mode it doesn't apply mutations
+- In strict it throws a `TypeError` if you try to apply mutations on a frozen object
+
+Be carefull, it isn't recursive. So our property commonNames isn't immutable.
+
+### Spread operator
+
+The spread operator syntax have been introduced in ES2015 for arrays and ES2018 for objects. It let's you copy all properties of an object in a new object literal.
+
+https://gist.github.com/frinyvonnick/61908a0e1bdcd97ba965570796c0c665
+
+For arrays, it copies all values of an array to a new array literal.
+
+https://gist.github.com/frinyvonnick/7ec4cd039fd79f16864546c51b18ab95
+
+Easily readable and unified between arrays and objects. It replaces nicely assign and concat. It is possible to spread multiple arrays and object in a same literal.
+
+## Why use immutability ?
+
+You found out how to make objects and arrays immutable with JavaScript but we didn't tell why using immutability is so necessary nowadays. As developper, we always search for a way to write more maintainable and readable code. Some paradigms as Functionnal Programming are focusing on this.
```

Comment:
`As developpers, we are always looking for a way to write more maintainable and readable code.`